libnavui-util/src/main/java/com/mapbox/navigation/ui/utils/internal/lifecycle/ViewLifecycleOwner.kt

```diff
@@ -0,0 +1,89 @@
+ * for example by collecting [Flow]s and drawing results in the nested views or on the map.
+ *
+ * In general, this lifecycle reflects the [hostingLifecycleOwner]'s [Lifecycle] whenever the `View` is attached
+ * but whenever the `View` is detached and we want to stop the coroutines tied to the lifecycle's scope,
+ * this lifecycle will transition to [Lifecycle.State.CREATED] (and invoke [Lifecycle.Event.ON_STOP])
+ * even if the hosting Activity or Fragment is still running.
+ *
+ * This lifecycle reaches [Lifecycle.State.DESTROYED] only when the [hostingLifecycleOwner] is destroyed.
+ *
+ * @see keepExecutingWhenStarted
+ */
+@UiThread
+class ViewLifecycleOwner(
+    private val hostingLifecycleOwner: LifecycleOwner
+) : LifecycleOwner {
+
+    private var isAttached = false
+
+    private val hostingLifecycleObserver = LifecycleEventObserver { _, event ->
+        val isAtLeastCreated = lifecycle.currentState.isAtLeast(Lifecycle.State.CREATED)
+        if (isAttached || (isAtLeastCreated && event == Lifecycle.Event.ON_DESTROY)) {
+            lifecycleRegistry.handleLifecycleEvent(event)
+        }
+    }
+    private val lifecycleRegistry = LifecycleRegistry(this)
+
+    init {
+        hostingLifecycleOwner.lifecycle.addObserver(hostingLifecycleObserver)
+    }
+
+    override fun getLifecycle(): Lifecycle = lifecycleRegistry
+
+    /**
+     * Call whenever [View.onAttachedToWindow] of the target [View] is called.
+     */
+    fun onAttachedToWindow() {
+        check(!isAttached) {
+            "onAttachedToWindow called without onDetachedFromWindow being called first"
+        }
+        lifecycleRegistry.currentState = hostingLifecycleOwner.lifecycle.currentState
+        isAttached = true
+    }
+
+    /**
+     * Call whenever [View.onDetachedFromWindow] of the target [View] is called.
+     */
+    fun onDetachedFromWindow() {
+        check(isAttached) {
+            "onDetachedFromWindow called without onAttachedToWindow being called first"
+        }
+        isAttached = false
+        if (hostingLifecycleOwner.lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED)) {
+            lifecycleRegistry.currentState = Lifecycle.State.CREATED
+        }
+    }
+
+    /**
+     * Launches a coroutine whenever this lifecycle reaches [Lifecycle.State.STARTED] or higher.
+     * If the lifecycle state triggers [Lifecycle.Event.ON_STOP] or beyond, the launched coroutines are canceled.
+     *
+     * If the state reaches [Lifecycle.State.STARTED] again, the coroutines are restarted.
+     * This is useful because the coroutines will keep being canceled and restarted
+     * as the target [View] is detached and re-attached to window.
+     *
+     * This function uses [repeatOnLifecycle] under the hood.
+     */
+    fun keepExecutingWhenStarted(block: suspend CoroutineScope.() -> Unit) = lifecycleScope.launch {
+        repeatOnLifecycle(Lifecycle.State.STARTED, block)
+    }
```

Comment:
IMO this does not belong in the `ViewLifecycleOwner`. It's essentially a helper extension and it will suggest we should continue adding more of these types of extensions into this class.

Another nitpick is it may not be a needed extension anyways.. the given extension is pretty good repeatOnLifecycle(STARTED) { .. }

Reply:
Do you suggest moving this to an extension function? Happy to do that, I was on the fence.

Reply:
oh i see, not just calling the function (my suggestion here https://github.com/mapbox/mapbox-navigation-android/pull/5332#discussion_r779774257 is not right)

But  only issue I have with it being a LifeycleOwner extension is that everyone who uses the SDK would inherit the helper function. But it should be scoped to our navigation sdk.

Reply:
That's why I initially added it to the class :D

That said, we don't expose the utils module by default (actually fixing it in this PR), so it wouldn't show up for regular consumers of the SDK. Someone would have to add the utils module explicitly to their build.

I'm gonna add an extension function for now. It's an `internal` package anyway, so we can change it if needed.

Reply:
Yeah agree, bigger issue is allowing UIComponents to take any `LifecycleOwner` so I prefer a global extension over that.

This actually isn't bad to have in `UIComponents` either. It may be good to communicate that a coroutine is being launched, maybe we'll fix this later with flow or live data.
```
lifecycleOwner.lifecycleScope.launch {
    lifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
```